### PR TITLE
TR-61: TS and Tests for rollup

### PR DESCRIPTION
**Pull request by shoumaw**
Migrated rollup to Typescript + added logic to test for metadata creation in the tests

### CODE REVIEW

File: public/jira/rollup/rollup.ts [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+  parentKeys: string[];
+}
+
+type ReportingHierarchyIssueOrRelease<CustomFields> =
+  IssueOrRelease<CustomFields> & {
+    reportingHierarchy: ReportingHierarchy;
+  };
+
+export interface RollupResponseItem<Data, Meta> {
+  metadata: Meta;
```
Comment (shoumaw) [inline public/jira/rollup/rollup.ts:27]: I removed the `RollupData<Data>` and `RollupMetadata<Meta>` fields
```
export type RollupData<D> = D & Record<string, any>;
export type RollupMetadata<M> = M & Record<string, any>;
```

And updated this
```
export interface RollupResponseItem<D, M> {
  metadata: RollupMetadata<M>;
  rollupData: RollupData<D>[];
}
```
to directly use `Data` and `Meta`
```
export interface RollupResponseItem<Data, Meta> {
  metadata: Meta;
  rollupData: Data[];
}
```
This seems to work fine after I made sure that the 3 generics are being passed down as needed from the main functions to all of the internal ones

Comment (shoumaw) [reply public/jira/rollup/rollup.ts:27]: There should not be any instance of `any` in the code

File: public/jira/rollup/rollup.test.ts [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+      { key: "i-7", parentKey: "m-3", myValue: 32 },
+    ],
+  ] as IssueOrRelease<{ myValue: number }>[][];
+
+  const results = rollupGroupedHierarchy<
+    { myValue: number },
+    { sumValue: number },
+    { hierarchyLevel: number; totalMyValue: number; itemCount: number }
+  >(groupedHierarchy.reverse(), {
+    createMetadataForHierarchyLevel(hierarchyLevel, issues) {
```
Comment (shoumaw) [inline public/jira/rollup/rollup.test.ts:108]: Passed down the callback that creates the Metadata to confirm that the types for it works as expected too

File: public/jira/rolledup/work-type/work-type.test.js [hunk 3]
```diff
@@ -1,132 +1,233 @@
...
-import { rollupDatesByWorkType, rollupWorkTypeDates } from './work-type.js'
-import { addReportingHierarchy, addChildrenFromGroupedHierarchy } from '../../rollup/rollup.js';
-
-
-function maintainOrderByKey(source, updated){
-    const keyToIndex = {};
-    for(let i = 0; i < source.length; i++) {
-        keyToIndex[source[i].key] = i;
+import { expect, test } from "vitest";
+import { rollupDatesByWorkType, rollupWorkTypeDates } from "./work-type.js";
```
Comment (DavidNic11) [inline public/jira/rolledup/work-type/work-type.test.js:3]: Should this fine be ts?

File: public/jira/rollup/rollup.ts (hunk 1 above)
Comment (DavidNic11) [reply public/jira/rollup/rollup.ts:27]: Is Data always going to be an array?

**Review by DavidNic11 — APPROVED** [review]

File: public/jira/rollup/rollup.ts (hunk 1 above)
Comment (shoumaw) [reply public/jira/rollup/rollup.ts:27]: Yes should be. 
All of the existing tests data expects rollupData to be an array. Including the tests of the custom rollup functions like `warning-issues`, `blocked-status-issues`, etc..

If a new rollup function was added that expects a different format, we can update it then. But for now type array aligns with the existing examples

File: public/jira/rolledup/work-type/work-type.test.js (hunk 3 above)
Comment (shoumaw) [reply public/jira/rolledup/work-type/work-type.test.js:3]: No this file `rolledup/work-type/work-type.js` is still a js file. Still to be migrated to typescript

**Pull request merged**